libeos-updater-util/flatpak.c

```diff
@@ -0,0 +1,1411 @@
+          g_set_error (error,
+                       EOS_UPDATER_ERROR,
+                       EOS_UPDATER_ERROR_MALFORMED_AUTOINSTALL_SPEC,
+                       "Expected '~locales' array to contain string values, was: %s",
+                       node_str);
+          return FALSE;
+        }
+
+      if (g_strv_contains ((const gchar * const *) supported_languages, string))
+        {
+          *is_filtered = TRUE;
+          return TRUE;
+        }
+    }
+
+  *is_filtered = FALSE;
+  return TRUE;
+}
+
+static gboolean
+locale_only_filter_applies (JsonNode  *filter_value,
+                                  gboolean  *is_filtered,
+                                  GError   **error)
+{
+  JsonArray *skip_locales_array = json_node_get_array (filter_value);
+  g_autoptr(GList) locale_nodes = NULL;
+  GList *iter = NULL;
+  g_auto(GStrv) supported_languages = NULL;
+
+  g_return_val_if_fail (is_filtered != NULL, FALSE);
+
+  if (!get_locales_list_from_flatpak_installation (&supported_languages,
+                                                   error))
+    return FALSE;
+
+  if (skip_locales_array == NULL)
+    {
+      g_autofree gchar *node_str = json_node_to_string (filter_value);
+      g_set_error (error,
+                   EOS_UPDATER_ERROR,
+                   EOS_UPDATER_ERROR_MALFORMED_AUTOINSTALL_SPEC,
+                   "Expected 'locales' filter to be an array, was: %s",
+                   node_str);
+      return FALSE;
+    }
+
+  locale_nodes = json_array_get_elements (skip_locales_array);
+
+  for (iter = locale_nodes; iter != NULL; iter = iter->next)
+    {
+      const gchar *string = json_node_get_string (iter->data);
+
+      if (string == NULL)
+        {
+          g_autofree gchar *node_str = json_node_to_string (filter_value);
+          g_set_error (error,
+                       EOS_UPDATER_ERROR,
+                       EOS_UPDATER_ERROR_MALFORMED_AUTOINSTALL_SPEC,
+                       "Expected 'locales' array to contain string values, was: %s",
+                       node_str);
+          return FALSE;
+        }
+
+      if (g_strv_contains ((const gchar * const *) supported_languages, string))
+        {
+          *is_filtered = FALSE;
+          return TRUE;
+        }
+    }
+
+  *is_filtered = TRUE;
+  return TRUE;
+}
+
+static gboolean
+action_filter_applies (JsonObject   *object,
+                       const gchar  *filter_key_name,
+                       gboolean     *is_filtered,
+                       GError      **error)
+{
+  JsonNode *filter_value = json_object_get_member (object, filter_key_name);
+
+  g_return_val_if_fail (is_filtered != NULL, FALSE);
+
+  if (g_strcmp0 (filter_key_name, "~architectures") == 0 &&
+      architecture_skip_filter_applies (filter_value, is_filtered, error))
+    return TRUE;
+  if (g_strcmp0 (filter_key_name, "architectures") == 0 &&
+           architecture_only_filter_applies (filter_value, is_filtered, error))
```

Comment:
Nitpick: Funky indentation here and in the other `if` statements in this function.